src/backend/joanie/badges/providers/obf.py

```diff
@@ -0,0 +1,468 @@
+    ):
+        """Override default requests.Session instantiation to handle authentication."""
+
+        super().__init__(*args, **kwargs)
+
+        self.api_root_url = "https://openbadgefactory.com"
+        self.api_version_prefix = "v1"
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.raise_for_status = raise_for_status
+        self.auth = self._get_auth
+        self.headers.update(
+            {
+                "Content-Type": "application/json",
+            }
+        )
+
+    @cached_property
+    def _access_token(self):
+        """Request OAuth2 access token from the API backend.
+
+        We cache this property to avoid regenerating an access token at each
+        API request. This access token has a limited validity (e.g. 10h) so we
+        try to regenerate it when the API response code is 403 (see the
+        `request` overridden method).
+
+        """
+
+        url = f"{self.api_version_prefix}/client/oauth2/token"
+        response = requests.post(
+            urljoin(self.api_root_url, url),
+            json={
+                "grant_type": "client_credentials",
+                "client_id": self.client_id,
+                "client_secret": self.client_secret,
+            },
+        )
+
+        if "access_token" not in response.json():
+            msg = "Cannot get an access token from the OBF server with provided credentials"
+            logger.error(msg)
+            logger.debug("Request status code was: %s", response.status_code)
+            logger.debug("Request URL was: %s", response.request.url)
+            logger.debug("Request body was: %s", response.request.body)
+            logger.debug("Response text was: %s", response.text)
+            raise AuthenticationError(msg)
+
+        return response.json().get("access_token")
+
+    @property
+    def _get_auth(self):
+        """Make access token generation dynamic."""
+
+        return OAuth2AccessToken(self._access_token)
+
+    def check_auth(self):
+        """Check OBF API credentials using the dedicated endpoint."""
+
+        url = f"/ping/{self.client_id}"
+        response = self.get(url)
+        if response.status_code != requests.codes.ok:  # pylint: disable=no-member
+            raise AuthenticationError("Invalid access token for OBF")
+        return response
+
+    @classmethod
+    def iter_json(cls, response: requests.Response) -> Iterable:
+        """Iterate over JSON lines serialization in API responses.
+
+        When multiple objects are returned by the API, they are streamed as
+        JSON lines instead of a JSON list, leading the response.json() method
+        to fail as it is expected a valid JSON list instead. We mitigate this
+        issue by forcing JSON serialization of each item in the response.
+        """
+
+        try:
+            json_ = response.json()
+            if isinstance(json_, list):
+                for item in json_:
+                    yield item
```

Comment:
```suggestion
                yield from json_
```

Maybe you can use a try/except pattern?
```
try:
    yield from json_
except ...:
    yield json_

Reply:
<img src="https://media0.giphy.com/media/Od0QRnzwRBYmDU3eEO/giphy.gif"/>

Reply:
In fact, as a dict is an iterator, I opt for the following:

```python
json_response = response.json()
if isinstance(json_response, list):
    yield from json_response
else:
    yield json_response
```